Indicate stock/target amount relations in stock menu

Noordfrees opened the pull request:
To allow getting an overview of economic shortages more easily, the ware icons in the stock menu now draw a red-tinged background if the stock is below the economy target setting, green if above, or yellow if equal; no tinge for wares without a target setting.
Since the stock menu shows only the sum of all wares in all economies, it also uses the sum of the target quantities of all non-trivial economies (trivial = has no warehouses) for this comparison.

Noordfrees committed:
Indicate stock/target amount relations in stock menu

Noordfrees committed:
2 files were automatically formatted.

frankystone commented:
I have tested this a bit, but i found it not very useful. Economy shortages can also be seen by watching the amount of wares. If one encounters the amount of a ware shrinks by watching the numbers, he also know that something is wrong with the economy. There is also the ware statistics, where one can see shortages.

But i am not against this proposal. Might be problematic for people with visual disorders.

Noordfrees commented:
For me watching the numbers is the hard part. I always have the stock window open in the screen corner to watch for shortages, but concentrating on those tiny numbers and evaluating whether they are high or low requires a lot of attention for me, which is why I would find this visual indicator very helpful there.

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

matthiakl commented:
I think this could be helpful, but also find the colored background a bit distracting sometimes (from the actual wares). Maybe make this feature toggleable with a button on the bottom? This would increase the window size, which might not be desirable, if you have the window constantly open...

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into stockmenu-compare-target

Noordfrees committed:
Add checkbox to toggle the new background colouring on/off

Noordfrees committed:
2 files were automatically formatted.

matthiakl reviewed (commented):
LGTM now. If @frankystone is ok with it, this can go in.

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

Noordfrees committed:
Merge branch 'master' into stockmenu-compare-target

frankystone commented:
I don't see a reason for this feature. Mainly because the it makes the ware statistics superfluous somehow and because the coloring has to be explained to the player and because it might be problematic for people with visual distractions. The text of the checkbox at the bottom is shown here as 'Monochrome' which is not very intuitive either.

I would vote for postponing this to the next version and mark this as 'under discussion', because there is no bug report or feature request open for implementing this.

Noordfrees commented:
> I don't see a reason for this feature. Mainly because the it makes the ware statistics superfluous somehow

There is no way of seeing currently whether your stock is above or below the target settings. The ware stats only show stock and production and consumption and differences, but no references to the target settings, and everything over time rather than momentary. So this is a new feature that adds previously unavailable information which can be used by the player, if desired, to finetune one's economy.

> and because the coloring has to be explained to the player and because it might be problematic for people with visual distractions. The text of the checkbox at the bottom is shown here as 'Monochrome' which is not very intuitive either.

Well how inverting the checkbox's behaviour so you get the old view by default, and this new feature is enabled only if you explicitly click it? The text can be changed to something more intuitive (we need to keep it short though as the window is pretty small). We can also add a tooltip to the checkbox to explain the colouring when the feature is switched on. Would this address all points?

> I would vote for postponing this to the next version and mark this as 'under discussion', because there is no bug report or feature request open for implementing this.

What would be the difference if I had opened an issue for this feature and *then* opened this PR? Of course I could just open an issue for this feature now (how about: "Bug: UI treats stock view and economy settings like unrelated entities even though they are logically closely linked") and link this PR to fix it ^^
(This is not meant to sound like I consider issues superfluous. But for me the point of an issue is to keep track of tasks that can, for whichever reason, not be performed immediately by the person who opens the issue; be it because they're not a dev or don't have the time or motivation to work on the thing immediately or whatever. If I *do* can and want to work on something – whether it's a bug or a feature – without delay I don't see the point of opening a new issue for it and instead create a PR to just *fix* it.)

frankystone commented:
> > I don't see a reason for this feature. Mainly because the it makes the ware statistics superfluous somehow
> 
> There is no way of seeing currently whether your stock is above or below the target settings. 

For me this is not very helpful. In my last game i set target setting for coal to 60 because at that time i didn't found any iron on the map. As a result every time the amount for coal in stock goes to 59 the background turns red, although there was enough coal. Yes, i could have decreased the target setting for coal after some time.

> The ware stats only show stock and production and consumption and differences, but no references to the target settings, and everything over time rather than momentary. So this is a new feature that adds previously unavailable information which can be used by the player, if desired, to finetune one's economy.

What about adding this information to the ware statistics?

> > I would vote for postponing this to the next version and mark this as 'under discussion', because there is no bug report or feature request open for implementing this.
> 
> What would be the difference if I had opened an issue for this feature and _then_ opened this PR? Of course I could just open an issue for this feature now (how about: "Bug: UI treats stock view and economy settings like unrelated entities even though they are logically closely linked") and link this PR to fix it ^^

Why do you feel angry? I have just told my view about this feature, not against your work.

If there were a feature request about this, the pros and cons could be discussed over there and not in this PR. Yes, of course then there might be the danger of endless discussions but in the end there will be a solution which will be the best for the game. And i think this should happen before one spend time to implement some thing.

Noordfrees, your contributions are really appreciated, but some times your style of discussion is really frustrating and one of the reasons why i do not want to contribute any more. Widelands is not a game for devs, it's a game for other people. If this consideration get lost, widelands will die.

Noordfrees commented:
If you feel angered or frustrated by my manner of debating I apologize, I did not mean to offend anyone. 

I feel that debates like this one have taken place multiple times before… Personally I invest the largest portion of my spare time in developing Widelands, and there is no bigger motivation killer for me than to have to debate about *small* changes for weeks or months. When there's a *big* change proposed which will necessarily affect and alter everyone's playing experience, then discussion is a necessity, and if there can be no consensus a proposal is rejected, however annoying this may be for the proposer.

But in this particular case, I genuinely do not understand what the argument actually is about. Initially, I changed the stock view to a different way. This change was not approved, which I can understand. So I added a checkbox to give everyone the choice whether the old or new display mode is used, and now I further proposed making the old design the default. So for you nothing would change except a new, never-used checkbox tucked at the edge of one window. I would get a useful new display mode to aid my economy management, to be switched on or off with a click. 

It is true that there are not many reason *for* adding this feature: There is only one, which is that at least one person (me) would consider it highly useful. Since you do not have to use this feature if you don't wish to, nor do anything at all to opt out, I genuinely cannot understand what then speaks *against* adding this feature though.

---

The reason why I want to put this indicator in the stock menu rather than the ware stats is this. The stock menu shows the momentary amounts of wares and workers in your economy. The ware preference window always shows information about ware amounts (and their changing) over a time span; and this information is displayed only for selected wares. If many wares are selected, there is no visual link between the wares and their graphs. And the graphs are all drawn at the same scale even though the orders of magnitude in which the wares are needed differ, e.g. 3 longswords are more than enough while 10 coal might be way too few, even though 10 is far higher than 3. It does not make sense to combine multiple types of information in a single graph. In the stock menu, on the other hand, each indicator is directly linked to the respective ware, and the window offers an overview of all wares at the same time. Only under this condition does the feature actually provide new useful information.

matthiakl commented:
I feel like this discussion becomes to big for such a small feature.

>>>I  don't see a reason for this feature. Mainly because the it makes the ware statistics superfluous somehow
>>
>> There is no way of seeing currently whether your stock is above or below the target settings.
>
> For me this is not very helpful.

You are discussing different play-styles here. I saw a very similar discussion over on #3981 regarding the Ctrl-click behavior. 
If it is not possible to please everyone with a new feature, but some highly request it, this is a good indicator for an optional feature, which it is with Nordfrees suggestion (and I proposed on the other one). So I vote for having the feature without changing the default design.

> I would vote for postponing this to the next version and mark this as 'under discussion', because there is no bug report or feature request open for implementing this.

Well, under discussion it already is. But there is no difference, whether it is discussed in an issue or a PR as long as the author accepts the risk to close it without merging.

> Noordfrees, your contributions are really appreciated, but some times your style of discussion is really frustrating [...]
> Widelands is not a game for devs, it's a game for other people.

I am overwhelmed with the time and contributions you put into this game, Nordfrees. Not every contributor is working on Widelands with the highest priority, so you shouldn't be upset when some of your PRs have a long waiting time.

Hopefully, I calmed down the discussion a bit, and we can continue to work towards a mutual approval.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into stockmenu-compare-target

Noordfrees committed:
Use old view as default mode, and add tooltip

Noordfrees committed:
2 files were automatically formatted.

frankystone commented:
As stated in my first post here, i am not against this :-)

In my opinion it would be useful if it shows the amount over some time.  Say the last few changes of amounts is stored in a stack and there is an indicator, which shows if the amounts grows or lowers. Something like arrows beside each ware: ↓↑

So one can see a trend, not only if the set target amount is reached or not.

Anyway, as said i am not against this.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into stockmenu-compare-target

Noordfrees committed:
Indicate stock trend over the last 5 minutes as '↑'/'='/'↓'

Noordfrees committed:
Saveloading fix

Noordfrees committed:
2 files were automatically formatted.

frankystone commented:
Thanks for being patient with me!

Tested it briefly and it works nicely now :-) 

The red arrow isn't much recognizable though. Can we have it brighter?
I am still not convinced about the colored backgrounds, it might get disturbing if they change in a short time, but if you think it helps, let's have it.
![stock_target_amounts](https://user-images.githubusercontent.com/6730556/105062162-96f3c480-5a7a-11eb-8a7c-c53a4f3a88e8.png)

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into stockmenu-compare-target

Noordfrees committed:
Brighter colours

Noordfrees committed:
2 files were automatically formatted.

frankystone commented:
I am fine now. Maybe choose RGB values 147, 100, 100 :-)

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into stockmenu-compare-target

Noordfrees committed:
Tweak red colour

Noordfrees commented:
This makes the arrow a bit hard to read against the black background so I changed it to 200,100,100.

BTW @matthiakl this change added some new required keys to the theme's `init.lua`

matthiakl reviewed (commented):
For me the colors are unobtrusive enough and fine now. 
The icon width only allows four characters, so the indicator should be skipped for amount >= 1000 (see inline comment).

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into stockmenu-compare-target

Noordfrees committed:
Address review

matthiakl reviewed (approved):
Let's have it :-)

matthiakl committed:
Merge branch 'master' into stockmenu-compare-target

matthiakl committed:
Merge branch 'master' into stockmenu-compare-target

hessenfarmer commented:
Maybe the comment might be too late, but I just stumbled over a potential logic error. The target of a ware decides whether it gets produced on stock so the total stocked amount can only get higher if there is a second productionsite (hunter for empire meat e.g.) or by a short overrun. so comparing totaö stocked versus target would be green on rare occassions wouldn't it?

Edit: solution might be lower then 50% of target --> red
lower then 75% of target --> neutral
else --> green

matthiakl commented:
> Maybe the comment might be too late, but I just stumbled over a potential logic error. The target of a ware decides whether it gets produced on stock so the total stocked amount can only get higher if there is a second productionsite (hunter for empire meat e.g.) or by a short overrun. so comparing totaö stocked versus target would be green on rare occassions wouldn't it?

The evaluation, whether a ware is produced is decided on stock amount, so wares on their way do not count. This normally results in a slight overshoot of the target and should make the green color visible enough.

hessenfarmer commented:
> 
> The evaluation, whether a ware is produced is decided on stock amount, so wares on their way do not count. This normally results in a slight overshoot of the target and should make the green color visible enough. 

Yes I know. but if they arrive and the colour is green production is stopped. Assuming a constant usage of the wares it will go to red in a couple of minutes and stay there for a longer time then it possibly could be green. Furthermore neutral color is even more seldom as it is just one small spot. Where the other colors are wide ranges.

matthiakl commented:
> Yes I know. but if they arrive and the colour is green production is stopped. Assuming a constant usage of the wares it will go to red in a couple of minutes and stay there for a longer time then it possibly could be green. Furthermore neutral color is even more seldom as it is just one small spot. Where the other colors are wide ranges.

Just give it a try for now, as I think over all wares it works quite well (tools with a low target get a neutral color more often, coal with a high target and large fluctuation less). This can be evaluated now, when some more people play with it. If you find the coloring really unbalanced, I'm sure we will find a way to fix it.
